autogluon/utils/tabular/ml/learner/default_learner.py

```diff
@@ -157,20 +160,26 @@ def general_data_processing(self, X: DataFrame, X_val: DataFrame, holdout_frac:
                 y_val = self.label_cleaner.transform(y_val)
         else:
             y_val = None
-
         # TODO: Move this up to top of data before removing data, this way our feature generator is better
         if X_val is not None:
             # Do this if working with SKLearn models, otherwise categorical features may perform very badly on the test set
             logger.log(15, 'Performing general data preprocessing with merged train & validation data, so validation performance may not accurately reflect performance on new test data')
-            X_super = pd.concat([X, X_val], ignore_index=True)
+            X_super = pd.concat([X, X_val, X_unlabeled], ignore_index=True)
             X_super = self.feature_generator.fit_transform(X_super, banned_features=self.submission_columns, drop_duplicates=False)
             X = X_super.head(len(X)).set_index(X.index)
             X_val = X_super.tail(len(X_val)).set_index(X_val.index)
+            if X_unlabeled is not None:
+                X_unlabeled = X_super.tail(len(X_unlabeled)).set_index(X_unlabeled.index)
             del X_super
         else:
-            X = self.feature_generator.fit_transform(X, banned_features=self.submission_columns, drop_duplicates=False)
+            X_super = pd.concat([X, X_unlabeled], ignore_index=True)
+            X_super = self.feature_generator.fit_transform(X_super, banned_features=self.submission_columns, drop_duplicates=False)
+            X = X_super.head(len(X)).set_index(X.index)
+            if X_unlabeled is not None:
+                X_unlabeled = X_super.tail(len(X_unlabeled)).set_index(X_unlabeled.index)
+            del X_super
```

Comment:
Seems like this code could be shortened to be less redundant.

Reply:
I think this can be done post-merge by core devs, the logic to make it short is a bit complex.